stage0/src/acpi.rs

```diff
@@ -0,0 +1,411 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// These data structures (and constants) are derived from qemu/hw/acpi/bios-linker-loader.c that
+// defines the interface.
+
+use crate::fw_cfg::FwCfg;
+use core::{
+    ffi::CStr,
+    fmt::{Debug, Formatter, Result as FmtResult},
+    mem::{zeroed, MaybeUninit},
+};
+use strum::FromRepr;
+
+// RSDP has to be within the first 1K of EBDA, so we treat it separately.
+// See Section 5.2.5 of the ACPI Specification, Release 6.5 for more details about RSDP, it's size,
+// structure, and where it should be placed.
```

Comment:
What for are we going to use ACPI ? IIUC it's a protocol to discover new devices?

Reply:
ACPI is many things; amongst others, yes, it's used to describe the hardware in the machine.

And that's exactly what we want to use it for: for example, if we use a MMIO (memory-mapped) virtio-vsock device in QEMU, we need to know what memory addresses were assigned to that device. We could go and just randomly start probing memory addresses, but a much better solution would be to just look at the device list in ACPI, find the descriptor for the device in the ACPI tables, and find the address for the device from the descriptor.

Further out in the future, ACPI contains details about CPUs in the system as well (which we need when we get to multiprocessing), and a ton of other stuff (like a more modern factility to trigger a shutdown).

Reply:
Who is populating the ACPI table with new devices?

Reply:
QEMU (or crosvm) provides us with the ACPI tables that contains information about hardware in the VM.

Crosvm writes the tables straight into guest memory; all of this code is about using the QEMU `fw_cfg` interface to read the ACPI tables and write them into the guest memory where the Restricted Kernel expects them to be.